### PR TITLE
os/: st7701 LCD driver with MIPI

### PR DESCRIPTION
**Pull request by namanjain7**

### CODE REVIEW

File: os/include/debug.h [hunk 1]
```diff
@@ -1146,6 +1146,30 @@ int get_errno(void);
 #define videoivdbg(...)
 #endif
 
+#ifdef CONFIG_DEBUG_XX_ERROR
```
Comment (edwakuwaku) [inline os/include/debug.h:1149]: Hi @namanjain7 , I think you misunderstood @sunghan-chang explanation, he is referring that when we add a new driver module, we will add log printing according to the module naming. For example, LCD, we use lcddbg, lcdvdbg.
Thus for MIPI, you should add like mipidbg, mipivdbg, etc.

File: os/drivers/mipidsi/mipi_dsi_packet.c [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+ ****************************************************************************/
+
+int mipi_dsi_create_packet(FAR struct mipi_dsi_packet *packet, FAR const struct mipi_dsi_msg *msg)
+{
+	if (packet == NULL) {
+		printf("packet is NULL \n");
+		return -EINVAL;
+	}
+	if (msg == NULL) {
+		printf("msg is NULL \n");
```
Comment (edwakuwaku) [inline os/drivers/mipidsi/mipi_dsi_packet.c:192]: And then for here, which is the driver module layer, you can use mipidbg for error logs, and mipivdbg for informational logs.
As discussed, printf is normally used from app layer.

File: os/include/tinyara/mipidsi/mipi_display.h [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+#define MIPI_DCS_ENTER_INVERT_MODE        0x21
+#define MIPI_DCS_SET_GAMMA_CURVE          0x26
+#define MIPI_DCS_SET_DISPLAY_OFF          0x28
+#define MIPI_DCS_SET_DISPLAY_ON           0x29
+#define MIPI_DCS_SET_COLUMN_ADDRESS       0x2a
+#define MIPI_DCS_SET_PAGE_ADDRESS         0x2b
+#define MIPI_DCS_WRITE_MEMORY_START       0x2c
+#define MIPI_DCS_WRITE_LUT                0x2d
+#define MIPI_DCS_READ_MEMORY_START        0x2e
+#define MIPI_DCS_SET_PARTIAL_ROWS         0x30	/* MIPI DCS 1.02 -
```
Comment (edwakuwaku) [inline os/include/tinyara/mipidsi/mipi_display.h:151]: Please check coding rule here, is it correct?

File: os/include/tinyara/mipidsi/mipi_dsi.h [hunk 4]
```diff
@@ -0,0 +1,864 @@
...
+
+/* Define cmd for dsi host */
+
+#define MIPIDSI_TRANSFER        _MIPIDSIIOC(6)
+
+/* Define msg flags */
+
+#define MIPI_DSI_MSG_REQ_ACK      (1 << 0)	/* Request ACK from peripheral */
+#define MIPI_DSI_MSG_USE_LPM      (1 << 1)	/* Use Low Power Mode to
+											 * transmit message */
```
Comment (edwakuwaku) [inline os/include/tinyara/mipidsi/mipi_dsi.h:71]: This file is having many misalignment, could you please check

File: apps/examples/lcd_test/example_lcd.c [hunk 5]
```diff
@@ -116,10 +116,10 @@ static void test_put_area(void)
 	fd = open(port, O_RDWR | O_SYNC, 0666);
 	if (fd < 0) {
 		printf("ERROR: Failed to open lcd port : %s error:%d\n", port, fd);
-		return -1;
+		return;
 	}
 	struct fb_videoinfo_s vinfo;
-	ioctl(fd, LCDDEVIO_GETVIDEOINFO, (unsigned long)(uintptr_t)&vinfo);
```
Comment (edwakuwaku) [inline apps/examples/lcd_test/example_lcd.c:122]: Please revert it to the original one.

File: os/drivers/mipidsi/Make.defs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+# language governing permissions and limitations under the License.
+#
+############################################################################
+
+ifeq ($(CONFIG_MIPI_DSI),y)
+
+CSRCS += mipi_dsi_device.c mipi_dsi_host.c mipi_dsi_packet.c mipi_dsi_device_driver.c mipi_dsi_host_driver.c
+
+ifeq ($(CONFIG_MIPI_DSI_DRIVER),y)
+	CSRCS += mipi_dsi_device_driver.c mipi_dsi_host_driver.c
```
Comment (edwakuwaku) [inline os/drivers/mipidsi/Make.defs:24]: Please check alignment

File: os/drivers/mipidsi/mipi_dsi_device.c [hunk 7]
```diff
@@ -0,0 +1,950 @@
...
+ *   OK on success or a negative error code on failure.
+ *
+ ****************************************************************************/
+
+int mipi_dsi_dcs_get_power_mode(FAR struct mipi_dsi_device *device, FAR uint8_t *mode)
+{
+	int ret;
+
+	ret = mipi_dsi_dcs_read(device, MIPI_DCS_GET_POWER_MODE, mode, sizeof(*mode));
+	if (ret <= 0) {
```
Comment (edwakuwaku) [inline os/drivers/mipidsi/mipi_dsi_device.c:548]:
```
	if (ret < 0) {
                return ret;
        }
	else if (ret == 0) {
		ret = -ENODATA;
	}
        else {
		return OK;
	}
```

File: os/Kconfig.debug [hunk 8]
```diff
@@ -1159,6 +1159,37 @@ config DEBUG_WLAN_API_INFO
...
 
+config DEBUG_MIPI
+	bool "MIPI Debug Feature"
+	default n
+	---help---
+		Enable MIPI debug feature.
+
+if DEBUG_MIPI
+
+config CONFIG_DEBUG_XX_ERROR
```
Comment (edwakuwaku) [inline os/Kconfig.debug:1170]: Please check comment, the xx should be mipi

File: os/include/tinyara/mipidsi/mipi_dsi.h (hunk 4 above)
Comment (namanjain7) [reply os/include/tinyara/mipidsi/mipi_dsi.h:71]: I ran the cformatter on all files as given in wiki page.
Didn't knew it make the code violate coding rules.

Comment (namanjain7) [reply os/include/tinyara/mipidsi/mipi_dsi.h:71]: I ran the cformatter on all files as given in wiki page.
Didn't knew it make the code violate coding rules.
I will fix this manually.

File: os/drivers/mipidsi/Make.defs (hunk 6 above)
Comment (namanjain7) [reply os/drivers/mipidsi/Make.defs:24]: I checked other Make.defs. Different modules have used different standards like tab, four spaces, two spaces.
I have removed the tab from here. Please let me know what is the correct way for Make.defs. As the wiki link for code standards isn't working.

File: os/drivers/mipidsi/mipi_dsi_device.c (hunk 7 above)
Comment (namanjain7) [reply os/drivers/mipidsi/mipi_dsi_device.c:548]: ok, will change this. 
For the else part, Mr. chang gave in the previous  comment to align like this ---> } else {

File: os/drivers/lcd/lcd_dev.c [hunk 9]
```diff
@@ -68,7 +68,6 @@
 #include <tinyara/fs/ioctl.h>
 #include <tinyara/lcd/lcd_dev.h>
 
-#define MAX_NO_PLANES 3
```
Comment (namanjain7) [inline os/drivers/lcd/lcd_dev.c:71]: I have added this macro in lcd_dev.h

File: os/include/debug.h (hunk 1 above)
Comment (namanjain7) [reply os/include/debug.h]: done

File: os/drivers/lcd/st7701.c [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+
+	struct st7701_config_s *config;
+};
+
+static struct st7701_dev_s g_lcdcdev;
+
+typedef struct lcm_setting_table {
+	u8 cmd;
+	u8 count;
+	u8 para_list[128];
```
Comment (edwakuwaku) [inline os/drivers/lcd/st7701.c:50]: In ST7701, the largest size for command list is 16, please revise the value here to:
`u8 para_list[16];`

**Comment by sunghan-chang** [comment]
@namanjain7 @anjana348 I've merged PR #6188 . Is this still necessary? If not, let's close.

**Comment by namanjain7** [comment]
Thr PR ##6188 has been merged. I have closed this PR.

**Pull request closed without merging**